enhancements/operator-dependency-resolution.md

```diff
@@ -0,0 +1,222 @@
+
+Internal details:
+Currently, the plan is to have `dependencies.yaml` as a centralized location for explicit dependency specification. However, we may plan to include that information in Dockerfile/manifest list so that information can be query without unpack the bundle image to see its content.
+
+#### Explicit Operator Dependency Validation
+
+As an operator author, I want to validate dependencies solution in the bundle. For example, I would like to make sure OLM will resolve and install all the dependencies I specify in the bundle correctly.
+* Allow operator authors to test dependency resolution for their operators before publishing
+
+Acceptance Criteria:
+I can use some toolings to list the operators that will be installed for my operator that will meet all of my dependency requirements that I specify in the bundle.
+
+Internal details:
+The proposal sat solver (resolver) will be in a separate repository in operator-framework and it can be imported into operator-registry and `opm` can use the resolver to perform the dependency resolution that OLM will do. Then, operator author can utilize the `opm` binary for testing and validation.
+
+#### Cluster Admin Dependency Observability
+
+As a cluster admin, I want to see the list of dependencies that are going to be installed when I inspect an operator.
+* Allow cluster admin to identify which operators that are going to be installed in order to satisfy the dependency requirement for a particular operator
+
+Acceptance Criteria:
+The cluster admin can see the dependency list in InstallPlan in the UI.
+
+Internal details:
+InstallPlan does contain a list of its components that it will install such as CSV, CRD and etc. In order to list operator dependency, it may require some changes in InstallPlan API to contain information about operator dependencies (an operator as a whole, not just operator components).
+
+### Implementation Details
+
+### Verify, Run and Test
+
+#### Explicit Operator Dependency Metadata
+
+The operator authors can specific explicit dependencies in `dependencies.yaml` in `/metadata` folder of bundle image.
+
+```
+$ tree test
+test
+├── manifests
+│   ├── operator.crd.yaml
+│   └── operator.v0.1.0.clusterserviceversion.yaml
+└── metadata
+    ├── dependencies.yaml
+    └── annotations.yaml
+```
+
+Test cases will be added to registry test suit to ensure `dependencies.yaml` is formatted correctly and being added to bundle image during the build process.
+
+#### Explicit Operator Dependency Validation
+
+A separate repo will be the home of the resolver (sat solver) and a new CI/test suit will be added. Then, a new command will be added to `opm` binary that can resolve the dependencies of a given operator bundle and index by using the resolver library in the new repo.
+
+```
+opm alpha resolve --images quay.io/test/test-operator:v0.0.1 --index quay.io/test/test-index:v0.0.1
```

Comment:
This is just a proposed command that I envision for how the command would theoretically look like. I don't see any issues with accepting more than one index. An index image is essentially a CatalogSource and there are often multiple CatalogSources in the cluster. So it makes sense to support multiple indexes/catsrc.

Reply:
I think we may need to expand on this some more if we want it to be a useful feature. Since we take into account things like existing operator groups, what namespace each catalog is in, etc. Do we want users to be able to model all of that? Or should we just allow it in the context of a set of indexes with equal value? Or maybe even just a single index?

Reply:
We can certainly add more tags/inputs to add more constraints for dependency resolution to work properly. We already have `InstallModes` information in CSV for the input bundle and in the indexes. The only missing constraints here is the namespace that this operator is gonna be installed which is translated into the OperatorGroup in that namespace. We can let user to provide a list of OperatorGroup information that represents the list of namespaces. For namespace of CatalogSource/index, I think if users are gonna provide the index, we should assume that CatalogSource/index is accessible in that particular namespace/list-of-namespaces.